zio-http/src/main/scala/zio/http/Flash.scala

```diff
@@ -0,0 +1,390 @@
+    }
+  }
+  private[http] object Message {
+    case class Notice[+A](a: A) extends Message[A, Nothing]
+    private[http] object Notice {
+      val name = "notice"
+    }
+    case class Alert[+B](b: B) extends Message[Nothing, B]
+    private[http] object Alert  {
+      val name = "alert"
+    }
+    private[http] case class Both[+A, +B](notice: Notice[A], alert: Alert[B]) extends Message[A, B]
+  }
+
+  /**
+   * `Flash.Backend` represents a flash-scope that is not cookie-based but
+   * instead uses an internal structure.
+   *
+   * Semantically it is identical to the cookie-based flash-scope (valid for a
+   * single request) but by using `Flash.Backend` we're not limited in size of
+   * the payload as in the cookie-based flash-scope. Still, the `Flash.Backend`
+   * uses a cookie but does not transport the payload with it but only an
+   * internal identifier.
+   */
+  trait Backend { self =>
+
+    /**
+     * Gets an `A` from the backend-based flash-scope or fails with a
+     * `Throwable`.
+     */
+    def flash[A](request: Request, flash: Flash[A]): IO[Throwable, A]
+
+    /**
+     * Gets an `A` from the backend-based flash-scope and provides a fallback.
+     */
+    def flashOrElse[A](request: Request, flash: Flash[A])(orElse: => A): UIO[A] =
+      self.flash(request, flash) <> ZIO.succeed(orElse)
+
+    /**
+     * Adds flash values to the backend-based flash-scope and returns a workflow
+     * with an updated `Response`.
+     */
+    def addFlash[A](response: Response, setter: Flash.Setter[A]): UIO[Response]
+
+    /**
+     * Optionally adds flash values to the backend-based flash-scope and returns
+     * a workflow with an updated `Response`.
+     */
+    def addFlash[A](response: Response, setterOpt: Option[Flash.Setter[A]]): UIO[Response] =
+      setterOpt.fold(ZIO.succeed(response))(self.addFlash(response, _))
+  }
+
+  object Backend {
+
+    private case class Impl(ref: Ref[Map[UUID, Map[String, String]]]) extends Backend {
+      override def flash[A](request: Request, flash: Flash[A]): IO[Throwable, A] =
+        for {
+          flashId <- ZIO.from(Flash.run(Flash.getUUID(flashIdName), request))
+          a       <- ref.modify { map =>
+            Flash.run(flash, map.get(flashId).getOrElse(Map.empty)) match {
+              case value @ Right(_) => value -> (map - flashId)
+              case value @ Left(_)  => value -> map
+            }
+          }.flatMap(ZIO.from(_))
+        } yield a
+
+      override def addFlash[A](response: Response, setter: Setter[A]): UIO[Response] = {
+        val map = Flash.Setter.run(setter, Map.empty)
+        for {
+          flashId       <- zio.Random.nextUUID
+          setterFlashId <- ref.update(in => in + (flashId -> map)).as(Flash.setValue(flashIdName, flashId))
+        } yield response.addFlash(setterFlashId)
+      }
+    }
+
+    val layer: ULayer[Backend] = ZLayer(Ref.make(Map.empty[UUID, Map[String, String]]).map(Impl.apply))
```

Comment:
I do not see the point of providing a flash backend based on a Map as most probably users' apps will be horizontally scaled 🤔

Reply:
Yes, you're right - when you're scaling the app that does not work anymore. In this case one needs to use an external data service. I think it should be possible to add another `Backend` with the given interface. 

Yes, calling it `inMemory` is a good idea!